From 0bfea7d66fe73f244e760218f0860b3ae3db5fa7 Mon Sep 17 00:00:00 2001
From: Vincent Donnefort <vdonnefort@google.com>
Date: Wed, 11 Mar 2026 16:55:57 +0000
Subject: [PATCH 199/223] ANDROID: KVM: arm64: Add
 __host_stage2_set_owner_complete()

Add a __host_stage2_set_owner_complete() function that must be called
after all __host_stage2_set_owner_locked() modifications are done. This
mostly intends to enable the host_donate_sglist_* functions. The
completion function is automatically called if HOST_SET_NO_COMPLETE is
unset.

For __pkvm_host_donate_sglist_guest() do the pvmfw loading after all the
sglist has been unmapped from both IOMMU and host stage-2.

Bug: 448672457
Signed-off-by: Vincent Donnefort <vdonnefort@google.com>
Cherrypick-From: https://android-review.googlesource.com/q/commit:8c75e994f559928748dacd062f86fc6865c2f581
Merged-In: I99601e1a8b71ac519415019f2d72d8d00443ad19
Bug: 520016024
Change-Id: I99601e1a8b71ac519415019f2d72d8d00443ad19
Signed-off-by: Chung-Kai (Michael) Mei <chungkai@google.com>
---
 arch/arm64/kvm/hyp/nvhe/mem_protect.c | 61 ++++++++++++++-------------
 1 file changed, 32 insertions(+), 29 deletions(-)

diff --git a/arch/arm64/kvm/hyp/nvhe/mem_protect.c b/arch/arm64/kvm/hyp/nvhe/mem_protect.c
index 94a508c235db..b1939cc42146 100644
--- a/arch/arm64/kvm/hyp/nvhe/mem_protect.c
+++ b/arch/arm64/kvm/hyp/nvhe/mem_protect.c
@@ -725,8 +725,19 @@ static void __host_update_page_state(phys_addr_t addr, u64 size, enum pkvm_page_
 enum host_set_page_state_flags {
 	HOST_SET_IS_MMIO		= BIT(0),
 	HOST_SET_NO_IOMMU_UPDATE	= BIT(1),
+	HOST_SET_NO_COMPLETE		= BIT(2), /* Skip __host_stage2_set_owner_complete() */
 };
 
+static void __host_stage2_set_owner_complete(u8 owner_id, enum host_set_page_state_flags flags)
+{
+	hyp_assert_lock_held(&host_mmu.lock);
+
+	if (flags & HOST_SET_NO_IOMMU_UPDATE)
+		return;
+
+	kvm_iommu_host_stage2_idmap_complete(owner_id == PKVM_ID_HOST);
+}
+
 static int __host_stage2_set_owner_locked(phys_addr_t addr, u64 size, u8 owner_id,
 					  enum pkvm_page_state nopage_state,
 					  enum host_set_page_state_flags flags)
@@ -751,11 +762,16 @@ static int __host_stage2_set_owner_locked(phys_addr_t addr, u64 size, u8 owner_i
 		return ret;
 
 	if (flags & HOST_SET_NO_IOMMU_UPDATE)
-		goto update_vmemmap;
+		goto complete;
 
 	prot = owner_id == PKVM_ID_HOST ? PKVM_HOST_MEM_PROT : 0;
 	kvm_iommu_host_stage2_idmap(addr, addr + size, prot);
-	kvm_iommu_host_stage2_idmap_complete(!!prot);
+
+complete:
+	if (flags & HOST_SET_NO_COMPLETE)
+		goto update_vmemmap;
+
+	__host_stage2_set_owner_complete(owner_id, flags);
 
 update_vmemmap:
 	if (flags & HOST_SET_IS_MMIO)
@@ -2269,24 +2285,13 @@ int __pkvm_host_split_guest(u64 gfn, u64 size, struct pkvm_hyp_vcpu *vcpu)
 	return ret;
 }
 
-static int __host_set_owner_guest(struct pkvm_hyp_vcpu *vcpu, u64 phys, u64 ipa,
+static void __host_set_owner_guest(struct pkvm_hyp_vcpu *vcpu, u64 phys, u64 ipa,
 				  size_t size, enum host_set_page_state_flags flags)
 {
-	struct pkvm_hyp_vm *vm = pkvm_hyp_vcpu_to_hyp_vm(vcpu);
 	u64 nr_pages = size >> PAGE_SHIFT;
-	int ret;
 
 	WARN_ON(__host_stage2_set_owner_locked(phys, size, PKVM_ID_GUEST, 0, flags));
 	psci_mem_protect_inc(nr_pages);
-	if (pkvm_ipa_range_has_pvmfw(vm, ipa, ipa + size)) {
-		ret = pkvm_load_pvmfw_pages(vm, ipa, phys, size);
-		if (WARN_ON(ret)) {
-			psci_mem_protect_dec(nr_pages);
-			return ret;
-		}
-	}
-
-	return 0;
 }
 
 int __pkvm_host_donate_guest(u64 pfn, u64 gfn, struct pkvm_hyp_vcpu *vcpu, u64 nr_pages)
@@ -2314,7 +2319,10 @@ int __pkvm_host_donate_guest(u64 pfn, u64 gfn, struct pkvm_hyp_vcpu *vcpu, u64 n
 	if (ret)
 		goto unlock;
 
-	WARN_ON(__host_set_owner_guest(vcpu, phys, ipa, size, 0));
+	__host_set_owner_guest(vcpu, phys, ipa, size, 0);
+
+	if (pkvm_ipa_range_has_pvmfw(vm, ipa, ipa + size))
+		WARN_ON(pkvm_load_pvmfw_pages(vm, ipa, phys, size));
 
 	prot = pkvm_mkstate(KVM_PGTABLE_PROT_RWX, PKVM_PAGE_OWNED);
 	WARN_ON(kvm_pgtable_stage2_map(&vm->pgt, ipa, size, phys, prot,
@@ -2389,18 +2397,14 @@ int __pkvm_host_donate_sglist_guest(struct pkvm_hyp_vcpu *vcpu)
 			goto unlock;
 	}
 
-	/*
-	 * Update the IOMMU outside of __host_set_owner_guest() so that
-	 * we can batch up the operations with a single call to
-	 * kvm_iommu_host_stage2_idmap_complete().
-	 */
 	for_each_hyp_ppage(ppage) {
 		size_t size = PAGE_SIZE << ppage->order;
 		u64 phys = hyp_pfn_to_phys(ppage->pfn);
+		u64 ipa = hyp_pfn_to_phys(ppage->gfn);
 
-		kvm_iommu_host_stage2_idmap(phys, phys + size, 0);
+		__host_set_owner_guest(vcpu, phys, ipa, size, HOST_SET_NO_COMPLETE);
 	}
-	kvm_iommu_host_stage2_idmap_complete(false);
+	__host_stage2_set_owner_complete(PKVM_ID_GUEST, 0);
 
 	for_each_hyp_ppage(ppage) {
 		size_t size = PAGE_SIZE << ppage->order;
@@ -2408,8 +2412,8 @@ int __pkvm_host_donate_sglist_guest(struct pkvm_hyp_vcpu *vcpu)
 		u64 ipa = hyp_pfn_to_phys(ppage->gfn);
 		enum kvm_pgtable_prot prot;
 
-		/* We already updated the IOMMU */
-		WARN_ON(__host_set_owner_guest(vcpu, phys, ipa, size, HOST_SET_NO_IOMMU_UPDATE));
+		if (pkvm_ipa_range_has_pvmfw(vm, ipa, ipa + size))
+			WARN_ON(pkvm_load_pvmfw_pages(vm, ipa, phys, size));
 
 		prot = pkvm_mkstate(KVM_PGTABLE_PROT_RWX, PKVM_PAGE_OWNED);
 		WARN_ON(kvm_pgtable_stage2_map(&vm->pgt, ipa, size, phys, prot,
@@ -2471,7 +2475,7 @@ int __pkvm_host_donate_sglist_hyp(struct pkvm_sglist_page *sglist, size_t nr_pag
 		if (ret) {
 			WARN_ON(ret != -ENOMEM);
 
-			kvm_iommu_host_stage2_idmap_complete(false);
+			__host_stage2_set_owner_complete(PKVM_ID_HYP, 0);
 
 			/* Rollback */
 			for (; p >= 0; p--) {
@@ -2485,15 +2489,14 @@ int __pkvm_host_donate_sglist_hyp(struct pkvm_sglist_page *sglist, size_t nr_pag
 			}
 			kvm_iommu_host_stage2_idmap_complete(true);
 
-			break;
+			goto unlock;
 		}
 
 		WARN_ON(__host_stage2_set_owner_locked(phys, size, PKVM_ID_HYP, 0,
-						       HOST_SET_NO_IOMMU_UPDATE));
-		kvm_iommu_host_stage2_idmap(phys, phys + size, 0);
+						       HOST_SET_NO_COMPLETE));
 	}
 
-	kvm_iommu_host_stage2_idmap_complete(false);
+	__host_stage2_set_owner_complete(PKVM_ID_HYP, 0);
 
 unlock:
 	hyp_unlock_component();
-- 
2.50.1

